### ISSUE TITLE
Should THCApply.cuh be using 'long long' instead of 'long'?

### ISSUE THREAD

**Issue by hughperkins** [torch/cutorch#191]
In THCApply.cuh, there is:

```
if (THC_canUse32BitIndexMath(state, a) &&
      THC_canUse32BitIndexMath(state, b) &&
      THC_canUse32BitIndexMath(state, c)) {
    TensorInfo<unsigned int> aInfo(state, a);
    TensorInfo<unsigned int> bInfo(state, b);
    TensorInfo<unsigned int> cInfo(state, c);

    HANDLE_A_CASE(unsigned int, aInfo.dims, bInfo.dims, cInfo.dims);
  } else {
    TensorInfo<unsigned long> aInfo(state, a);
    TensorInfo<unsigned long> bInfo(state, b);
    TensorInfo<unsigned long> cInfo(state, c);
```

Per https://software.intel.com/en-us/articles/size-of-long-integer-type-on-different-architecture-and-os/ , a 'long' is basically a 32-bit int.  This article suggests that if one wants to obtain 32-bit and 64-bit integers, somewhat pseudo-portably, currently one can use the following types:

* 32-bit integer: int
* 64-bit integer: long long

**Comment by hughperkins** [comment]
(Note that I'm starting to move cltorch to using int64 and uint64 typedefs, rather than long and long long, https://github.com/hughperkins/cltorch/commit/6b06f6ffdf20c741bed591cc12ee7c05115826cc , which apparently was sufficient to fix a mac build issue https://github.com/hughperkins/cltorch/issues/3#issuecomment-118125371 )

**Comment by dominikgrewe** [comment]
I agree. Using `long` is not great. Torch uses it all over the place, probably with the assumption that it's 64-bit (which I believe is true for most modern, i.e. 64-bit, Linux and Mac systems). I'd love to see `int64` and `uint64` instead.

**Comment by hughperkins** [comment]
Ok, sounds good.  I started attempting this, got a bit more than I could chew.  Some things to ponder:
- like const-casts, it sort of 'infects' everything, since array of  `long long` is not compatible with array of `long`
- since trying to change the whole world is inevitably doomed to failure, I'm thinking that maybe the following function could be used at the boundary between `int64` areas and `long` areas:
```
long *int64ToLong( int64 *val ) {
  if( sizeof(long) == sizeof(int64) ) {
    return (long *)val;
  } else {
    throw std::runtime_error("Incompatible types, 32-bit vs 64-bit int array");
  }
}
```
- anywhere that a long is actually 64-bit, ti works just fine
- anywhere that it's not, do we actually want it to start pretending they are?  perhaps better to crash in a somewhat-controlled fashion, and fix it?

that's the first point.

second point is, from the point of view of the user-interface into torch, I think it's well established that `Long` means 64-bit?  And this is pretty standard, eg java http://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html .  its only c/c++ that has some ambiguity about it.

So, maybe LongStorage nad LongTensor stays LongStorage and LongTensor, including when they're used as c/c++ functoin names, structs, classes etc?

Only thing that changes is the underlying c/c++ types "tighten up" a bit, so they dont behave in differnet ways, depending on whether one is on Windows, mac, 32-bit/64-bit linux, etc?

Oh yeah, and one other thing: might need to start in Torch first :-P  eg in cutorch or cltorch, Tensor.c line 1070 calls into THFile_writeLongRaw, passing an array of longs.  However, using the somewhat-hacky function above, this could be manageable, without first modifying Torch?

**Comment by hughperkins** [comment]
Hmmm: maybe the first step of having a smooth migraiton of both torch and cu/cltorch would be to ensure that cu/cltorhc can accept IntTensors anywhere they can accept LongTensors.  Then, I think migrating torch itself is fairly smooth, posisbly.

In reality, I'm probalby just going to migrate stuff across as I hit weird build problems and stuff, no major hurry :-)

**Comment by wickedfoo** [comment]
CUDA uses `long long` for 64 bit integers so we should probably use that. What size `long` or `long long` uses is a compiler-dependent type; I think the only guarantee is that sizeof(int) <= sizeof(long) <= sizeof(long long).

To be truly portable one should use the C99 types int32_t and int64_t:

http://en.cppreference.com/w/c/types/integer

**Comment by wickedfoo** [comment]
However, CUDA by default (invoking nvcc from the command line, with no args) has sizeof(long) == sizeof(long long) == 8 in whatever configuration that uses but they are different types, hence some typecasts floating around.

**Comment by wickedfoo** [comment]
The CUDA APIs tend to use `long long`.

**Comment by hughperkins** [comment]
`long long` would work.  I didnt know about int32_t and int64_t.  Since C++ is not a superset of C99, but a separate fork essentially, I'm wondering whether MSVC C++ supports these?

**Comment by wickedfoo** [comment]
it's in C++ too, 11 I think officially but I think also in most 98 or 03
installations.

**Comment by wickedfoo** [comment]
@soumith I think this was causing problems for your ARM port thing too.

If we don't want to go to the standard int32_t/uint32_t/int64_t/uint64_t, I'd vote for going to a typedef header to define our own types in cutorch, and changing all instances of types in cutorch to point to that.

torch7 likely needs to be addressed in a similar way as well. But cutorch can at least change independently.

**Comment by hughperkins** [comment]
> But cutorch can at least change independently.

with the proviso of trying not to go down the rabbit hole when you hit https://github.com/torch/cutorch/blob/df463695f0cd387736917e02abaafa63c00bbed3/torch/generic/Tensor.c#L1134-L1137 :
```
  THFile_writeIntScalar(file, tensor->nDimension);
  THFile_writeLongRaw(file, tensor->size, tensor->nDimension);
  THFile_writeLongRaw(file, tensor->stride, tensor->nDimension);
  THFile_writeLongScalar(file, tensor->storageOffset+1); /* to respect Lua convention */
```
I got quite far with my initial attempt last year, but sort of gave up trying to solve this bit, I think, seemed like it would need a bunch of coordination, between torch and cutorch, to fix this, perhaps?
